controller/src/test/java/org/jboss/as/controller/test/ReadFeatureDescriptionTestCase.java

```diff
@@ -0,0 +1,623 @@
+import static org.jboss.as.controller.operations.global.ReadFeatureDescriptionHandler.OPERATION_PARAMETERS_MAPPING;
+import static org.jboss.as.controller.operations.global.ReadFeatureDescriptionHandler.PARAMETERS;
+import static org.jboss.as.controller.operations.global.ReadFeatureDescriptionHandler.REFERENCES;
+
+
+public class ReadFeatureDescriptionTestCase extends AbstractControllerTestBase {
+
+    private static final String TEST_EXTENSION = "org.wildfly.testextension";
+    private static final String TEST_SUBSYSTEM = "testsubsystem";
+    private static final String RESOURCE = "resource";
+    private static final String MAIN_RESOURCE = "main-resource";
+    private static final String SPECIAL_NAMES_RESOURCE = "special-names-resource";
+    private static final String REFERENCING_RESOURCE = "referencing-resource";
+    private static final String RUNTIME_RESOURCE = "runtime-resource";
+    private static final String NON_FEATURE_RESOURCE = "non-feature-resource";
+    private static final String TEST = "test";
+    private static final String MAIN_RESOURCE_CAPABILITY_NAME = "main-resource-capability";
+    private static final String ROOT_CAPABILITY_NAME = "root-capability";
+
+    private static final OperationStepHandler WRITE_HANDLER = new ModelOnlyWriteAttributeHandler();
+
+    private ManagementResourceRegistration registration;
+
+    @Override
+    public void setupController() throws InterruptedException {
+        super.setupController();
+        // register read-feature op
+        // can't do this in #initModel() because `capabilityRegistry` is not available at that stage
+        registration.registerOperationHandler(ReadFeatureDescriptionHandler.DEFINITION,
+                ReadFeatureDescriptionHandler.getInstance(capabilityRegistry), true);
+    }
+
+    /**
+     * Reads subsystem that has sub-resources, but no attributes. Subsystem provides a capability.
+     *
+     * Expectations:
+     *
+     * - feature-id param,
+     * - param for extension which contains this subsystem,
+     * - provided capability is listed in 'provides',
+     * - 'refs' contain {feature = "extension", include = true},
+     * - packages contain "[extension name].main".
+     */
+    @Test
+    public void testSubsystem() throws OperationFailedException {
+        PathAddress address = PathAddress.pathAddress(SUBSYSTEM, TEST_SUBSYSTEM);
+        ModelNode result = readFeature(address);
+
+        ModelNode feature = result.require(FEATURE);
+        Assert.assertEquals(serializeAddress(address), feature.require(NAME).asString());
+
+        // check params
+        Map<String, ModelNode> params = extractParamsToMap(feature);
+        Assert.assertEquals(2, params.size());
+        assertFeatureIdParam(SUBSYSTEM, TEST_SUBSYSTEM, params); // path parameter
+        assertParamWithDefault(EXTENSION, TEST_EXTENSION, params); // referencing extension
+
+        // check provides
+        ModelNode provides = feature.require("provides");
```

Comment:
likewise if some constants are missing that are used in the model of the `read-feature` operation, they should be introduced in `org.jboss.as.controller.descriptions.ModelDescriptionConstants` and used in org.jboss.as.controller.operations.global.ReadFeatureDescriptionHandler